streams/src/main/java/org/apache/kafka/streams/processor/internals/RackAwareTaskAssignor.java

```diff
@@ -0,0 +1,164 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.TopicPartitionInfo;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.internals.TopologyMetadata.Subtopology;
+import org.apache.kafka.streams.processor.internals.assignment.AssignorConfiguration.AssignmentConfigs;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RackAwareTaskAssignor {
+    private static final Logger log = LoggerFactory.getLogger(RackAwareTaskAssignor.class);
+
+    private final Cluster fullMetadata;
+    private final Map<TaskId, Set<TopicPartition>> partitionsForTask;
+    private final Map<UUID, Map<String, Optional<String>>> processRacks;
+    private final AssignmentConfigs assignmentConfigs;
+    private final Map<TopicPartition, Set<String>> racksForPartition;
+    private final InternalTopicManager internalTopicManager;
+    private Boolean canEnableForActive;
+
+    public RackAwareTaskAssignor(final Cluster fullMetadata,
+                                 final Map<TaskId, Set<TopicPartition>> partitionsForTask,
+                                 final Map<Subtopology, Set<TaskId>> tasksForTopicGroup,
+                                 final Map<UUID, Map<String, Optional<String>>> processRacks,
+                                 final InternalTopicManager internalTopicManager,
+                                 final AssignmentConfigs assignmentConfigs) {
+        this.fullMetadata = fullMetadata;
+        this.partitionsForTask = partitionsForTask;
+        this.processRacks = processRacks;
+        this.internalTopicManager = internalTopicManager;
+        this.assignmentConfigs = assignmentConfigs;
+        this.racksForPartition = new HashMap<>();
+    }
+
+  public synchronized boolean canEnableForActive() {
+      if (canEnableForActive != null) {
+          return canEnableForActive;
+      }
+
+      /*
+      TODO: enable this after we add the config
+      if (StreamsConfig.RACK_AWARE_ASSSIGNMENT_STRATEGY_NONE.equals(assignmentConfigs.rackAwareAssignmentStrategy)) {
+          canEnableForActive = false;
+          return false;
+      }
+       */
+
+      if (!validateClientRack()) {
+          canEnableForActive = false;
+          return false;
+      }
+
+      canEnableForActive = validateTopicPartitionRack();
+      return canEnableForActive;
+  }
+
+  public boolean canEnableForStandby() {
+      // TODO
+      return false;
+  }
+
+  private boolean validateTopicPartitionRack() {
+      // Make sure rackId exist for all TopicPartitions needed
```

Comment:
The reason I want to enforce all topic partition racks exist to enable the assignor is to avoid too much assignment changes if some racks appear or disappear later somehow.